### ISSUE TITLE
Notifications GraphQL requires `timestamp` field is non-nullable

### ISSUE THREAD

**Issue by josias-r** [directus/directus#14100]
### Describe the Bug

**TL;DR**
1. Notifications GraphQL API expects wrong types and doesn't work at all
2. Notifications REST API requires EMAIL to be set-up? Not mentioned in the docs + why so? Notifications should be able to be just in-application without an email tied to it.
3. Notifications REST API can be abused even without ANY auth header
4. Notifications REST API: the Notification created will link to an invalid url in the app (`/admin/content/null/null`)
---

I tried following the [docs for the notifications API](https://docs.directus.io/reference/system/notifications/#notifications) to post a message to a user.

According to that doc's example, all that is needed to POST a notification are the `recipient` and `subject`.

First I tried it using GraphQL with the following query:
```ts
const systemMutation = gql`
  mutation createNotification(
    $USER: ID!
    $MESSAGE: String!
  ) {
    create_notifications_item(
      data: {
        recipient: $USER
        subject: $MESSAGE
      }
    ) {
      id
    }
  }
`;
const sysemMutationResult: {
  create_notifications_item: {
    data: {
      id: string;
    };
  };
} = await graphQLSystemClient.request(systemMutation, {
  USER: notification.user,
  MESSAGE: notification.message,
});

```

And I received the first odd error:
> `timestamp` (type `String!`) is supposedly required

I thought that's weird, but I guess I'll just generate one myself and add it to the `data`.

After adding `timestamp` to the `data` object, I received the next odd error:
> `$USER` is supposed to be of type `create_directus_users_input!` and not `ID!`

That did not make any sense to me. I tried playing around with data for that type, but `create_directus_users_input` seems to expect a full user object, not just the primary id (which makes no sense, of course, I don't want to create a new user)

So I gave up on using the GraphQL endpoint and thought, that maybe the REST API will work.

```ts
const response = await fetch(`${process.env.DIRECTUS_REST_ENDPOINT}/notifications`, {
  method: "POST",
  headers: {
    "Content-Type": "application/json",
    Authorization: "Bearer XXXXX",
    Accept: "application/json", // required
  },
  body: JSON.stringify({
    recipient: "xxxxx-xxxx-xxxx-xxxx-xxxxxxxx",
    subject: ":)))",
  }),
});
```
The REST API actually kinda worked but still had some weird quirks:
1. The EMAIL needed to be setup, otherwise there was an error 500 (`Error: Sendmail exited with code 1`)
2. Even if **no** `Authorization` header was provided **a Mail was still sent out** (after configuring mail)
   - > Seems like quite the security issue, but I think it's not THAT big of a deal, because someone trying to abuse this would still need to know userIDs first I think
3. The Notification created will link to an invalid url in the app (`/admin/content/null/null`)

> I used the endpoints `http://localhost:8055/graphql/system` and `http://localhost:8055/notifications`
> The Bearer token used had admin privileges

### To Reproduce

Run similar queries to the ones mentioned above

docker-compose.yml used locally:
```yml
version: '3'
services:
  database:
    container_name: database
    image: postgis/postgis:latest
    volumes:
      - ./data/database:/var/lib/postgresql/data
    networks:
      - directus
    environment:
      POSTGRES_USER: 'postgres'
      POSTGRES_PASSWORD: 'postgres'
      POSTGRES_DB: 'postgres'

  cache:
    container_name: cache
    image: redis:latest
    networks:
      - directus

  directus:
    container_name: directus
    image: directus/directus:latest
    ports:
      - 8055:8055
    volumes:
      - ./uploads:/directus/uploads
      - ./extensions:/directus/extensions
    networks:
      - directus
    depends_on:
      - cache
      - database
    environment:
      KEY: 'xxx'
      SECRET: 'xxxx'

      DB_CLIENT: 'pg'
      DB_HOST: 'database'
      DB_PORT: '5432'
      DB_DATABASE: 'postgres'
      DB_USER: 'postgres'
      DB_PASSWORD: 'postgres'

      CACHE_ENABLED: 'true'
      CACHE_STORE: 'redis'
      CACHE_REDIS: 'redis://cache:6379'

      ADMIN_EMAIL: 'xxx@example.ch'
      ADMIN_PASSWORD: 'xxx'

      EMAIL_TRANSPORT: 'smtp'
      EMAIL_SMTP_HOST: 'mail.example.ch'
      EMAIL_SMTP_PORT: '999'
      EMAIL_SMTP_USER: 'acc@example.ch'
      EMAIL_SMTP_PASSWORD: 'xxx'

      EXTENSIONS_AUTO_RELOAD: 'true'

      # Make sure to set this in production
      # (see https://docs.directus.io/self-hosted/config-options/#general)
      # PUBLIC_URL: 'https://directus.example.com'

networks:
  directus:

```

### Errors Shown

GraphQL error when trying to create a notifications item without a timestamp in the `data`
```
Field "create_directus_notifications_input.timestamp" of required type "Date!" was not provided.
```
GraphQL error when trying to create a notifications item with the recipient being type `ID!`
```
Variable "$USER" of type "ID!" used in position expecting type "create_directus_users_input"
```
GraphQL error that illustrates that the expected recipient is a full object and not just the primary key
```
Expected type "create_directus_users_input" to be an object.
```
Console error when REST API is called without email being set-up
```
Error: Sendmail exited with code 1
```

### What version of Directus are you using?

9.12.2

### What version of Node.js are you using?

v16.14.0

### What database are you using?

Postgres 14.3

### What browser are you using?

Chrome

### How are you deploying Directus?

docker-compose locally

**Comment by azrikahar** [comment]
> 1. Notifications GraphQL API expects wrong types and doesn't work at all

For the `timestamp` being required, can confirm it's an issue. Seems to be due to the `notNullable()` here:

https://github.com/directus/directus/blob/6da4bd4a76b80284a07cc391b4fbd7df967e0dcb/api/src/database/migrations/20211118A-add-notifications.ts#L6

For the recipient error, the correct input would be `recipient: {id: $USER}` instead of `recipient: $USER`. However if you attempt to do so in this create mutation, it'll request you to add status and provider since they are required fields in directus_users. This is being tracked in #7237.

> 2. Notifications REST API requires EMAIL to be set-up? Not mentioned in the docs + why so? Notifications should be able to be just in-application without an email tied to it.

It doesn't technically requires it, but as users have "Email Notifications" enabled by default, hence it indirectly causes a dependence on email being setup for it to work for users that have it enabled (you should be able to test this by disabling "Email Notifications" for the recepient). That said, it definitely shouldn't be tied and this is currently being tracked in #13974.

> 3. Notifications REST API can be abused even without ANY auth header

This seems odd. It should not be accessible without having an auth header (or access token) by default. I do get 403 Forbidden when I attempt to do so:

![chrome_Q11WTizeVX](https://user-images.githubusercontent.com/42867097/175844084-173e1f1e-37bd-4122-81b4-eb0c8e460cc7.png)

Were you granting the create permissions on Directus Notifications to the Public role by any chance?

> 4. Notifications REST API: the Notification created will link to an invalid url in the app (`/admin/content/null/null`)

Just to provide a bit of context, the MVP of notifications was set out to complement the comment system, hence the it was assuming notifications are from a comment that links to `collection-name/item-name` (`null/null` in your case). This is being tracked in #13727.

**Comment by josias-r** [comment]
> For the recipient error, the correct input would be recipient: {id: $USER} instead of recipient: $USER. However if you attempt to do so in this create mutation, it'll request you to add status and provider since they are required fields in directus_users. This is being tracked in https://github.com/directus/directus/issues/7237.

Ah awesome! I tried it as an object too liked you mentioned and ran into exactly that problem. The possibility to "update" via a nested query would be amazing, I already had a few use-cases for it

> This seems odd. It should not be accessible without having an auth header (or access token) by default. I do get 403 Forbidden when I attempt to do so:

Letme gather some reproduction instructions for this. But I was using directus' default permissions. I think I received a 403 too, but the email was still sent in the background

> Just to provide a bit of context, the MVP of notifications was set out to complement the comment system, hence the it was assuming notifications are from a comment that links to collection-name/item-name (null/null in your case). This is being tracked in https://github.com/directus/directus/issues/13727.

I've figured that out in the meantime too and I think this covers most use-cases anyways. I also think I saw something in the DB on the user level to opt-out of getting emails on top of the in-app notifications. Would be cool if that default behaviour could also be decided during the API call, something like `shouldSendEmail?: boolean`

**Comment by josias-r** [comment]
> Letme gather some reproduction instructions for this.
Okay, here you go:
Use the following docker-compose and run it:
```
version: '3'
services:
  database:
    image: postgis/postgis:14-3.2
    volumes:
      - directus-postgis-14:/var/lib/postgresql/data
    environment:
      POSTGRES_USER: 'postgres'
      POSTGRES_PASSWORD: 'postgres'
      POSTGRES_DB: 'postgres'
    networks:
      - directus
    ports:
      - 5432:5432
  cache:
    image: redis:7.0.2
    networks:
      - directus
  directus:
    image: directus/directus:9.12.2
    ports:
      - 8055:8055
    networks:
      - directus
    depends_on:
      - cache
      - database
    environment:
      KEY: '11111111-2222-3333-4444-555555555555'
      SECRET: '66666666-7777-8888-9999-000000000000'

      DB_CLIENT: 'pg'
      DB_HOST: 'database'
      DB_PORT: '5432'
      DB_DATABASE: 'postgres'
      DB_USER: 'postgres'
      DB_PASSWORD: 'postgres'

      CACHE_ENABLED: 'true'
      CACHE_STORE: 'redis'
      CACHE_REDIS: 'redis://cache:6379'
      # EMAIL_TRANSPORT: 'smtp'
      # EMAIL_SMTP_HOST: ${EMAIL_SMTP_HOST}
      # EMAIL_SMTP_PORT: ${EMAIL_SMTP_PORT}
      # EMAIL_SMTP_USER: ${EMAIL_SMTP_USER}
      # EMAIL_FROM: ${EMAIL_FROM}
      # EMAIL_SMTP_PASSWORD: ${EMAIL_SMTP_PASSWORD}

      # Make sure to set this in production
      # (see https://docs.directus.io/self-hosted/config-options/#general)
      # PUBLIC_URL: 'https://directus.example.com'
volumes:
  directus-postgis-14:
networks:
  directus:
```
- Check the console for the admin credentials.
- Log in.
- Change the email to your own existing email (or just set the env var in the beginning)
- Copy the admin user's ID hash
- Send a POST to `http://localhost:8055/notifications`:
   -  ```
       {
          "recipient": "<PAST-ADMIN-USER-HASH>",
           "subject": "Hi there!"
       }
      ```
- You will receive a 500 and in the console see the following:
   - ![Screenshot 2022-06-27 at 09 07 13](https://user-images.githubusercontent.com/11424820/175881402-bd6dad4c-5d3e-4da4-8e5d-f931242048ae.png)
- Shut down the docker-compose process
- Change the commented out `EMAIL_*` env vars in the docker-compose file to a real SMTP server+account
- Start up docker again
- Send a POST to `http://localhost:8055/notifications` again.
- You will receive: 
   - ![Screenshot 2022-06-27 at 09 09 55](https://user-images.githubusercontent.com/11424820/175881827-47344726-9735-45e1-b121-b8003b144d33.png)
- **But you will also receive an email in your inbox**, even though there was a 403 -> Check your inbox

**Comment by azrikahar** [comment]
@josias-r Thanks again for pointing out email being sent even when the notification didn't have the necessary permission! It has been addressed in https://github.com/directus/directus/pull/14131#discussion_r907809270 👍

**Comment by azrikahar** [comment]
Just as an additional recap and heads up:

- Although the linked PR #14323 will resolve the `timestamp` field being required error, we will still face issue on GraphQL creation as it'll request us to add `status` and `provider` since they are required fields in `directus_users`. This is because it is treating it as _create_ user rather than associate the user instead. This is being tracked in #7237.

- The sending of email despite notification not created has been resolved in #14131.

- Currently the notifications created via Flows will link to an invalid URL in the app (`/admin/content/null/null`). This is being tracked in #13727.

With the points above in mind, #14323 will proceed to close this issue when it gets merged as the other issues are tracked elsewhere separately. 👍

**Comment by josias-r** [comment]
Awesome, you guys are doing amazing work 🚀
